src/Controller/ControllerFactory.php

```diff
@@ -78,13 +78,37 @@ public function create(ServerRequestInterface $request): Controller
         if ($reflection->isAbstract()) {
             throw $this->missingController($request);
         }
+        $this->container->addShared(
+            ComponentRegistry::class,
+            new ComponentRegistry(null, $this->container)
+        );
 
         // Get the controller from the container if defined.
         // The request is in the container by default.
         if ($this->container->has($className)) {
             $controller = $this->container->get($className);
         } else {
-            $controller = $reflection->newInstance($request);
+            $components = $this->container->get(ComponentRegistry::class);
+            $constructor = $reflection->getConstructor();
+            assert($constructor !== null);
+            $hasComponents = false;
+            foreach ($constructor->getParameters() as $parameter) {
+                $paramType = $parameter->getType();
+                // TODO: In a future minor release it would be good to start requiring the components parameter
```

Comment:
People won't know they need to upgrade the constructor signature if we don't show a warning.

Reply:
Yes, do we want to add a deprecation for this now, or in a future release? I know I usually regret letting deprecation TODOs accumulate, but also don't want to bombard developers on the first point release.

Reply:
I guess we can punt the message to 5.2.

Reply:
That was my thinking too.